### PR TITLE
Remove `object_id` from `Subscriptions` and update Subscription + Notification Tests

### PR DESCRIPTION
**Pull request by timolegros**
## Link to Issue
Closes: #4366 

### Client Side
- Clean-up `NotificationSubscription` client model:
  - Removed ambiguity between a full chain/thread/comment instance and just the id -> always have a full instance.
- Refactored sections of Notifications client controller
  - Added a `findNotificationSubscription` method which matches some search parameters to a NotificationSubscription. The parameter for this function uses the new `SubUniqueData` type which maps a `NotificationCategory` to an object containing the minimum necessary data to uniquely identify a `NotificationSubscription`
  - Removed public access to the NotificationSubscription array. This is to force other parts of the application to use `findNotificationSubscription`.
  - Updated `parseNotifications` method so that it doesn't require a subscription
  - Added some utility-getter methods to retrieve discussion and chain event subscriptions separately
  - Updated the subscribe function to use the `SubUniqueData` type. Also updated all `app.user.notifications.subscribe` calls.
- Clean-up `Notification` client model:
  - Removed the need for a full subscription associated to each notification.
- Refactored all `app.user.notifications.subscriptions.find` instances to use `app.user.notifications.findNotificationSubscription` method instead.
- Deleted `packages/commonwealth/client/scripts/helpers/notifications.ts` since it is unused.
- Renamed IPostNotificationData to IForumNotificationData for clarity.
### Server Side
- Created an interface for every NotificationCategory `notification_data`. This means all notification data is typed. Since I had to edit every single `notification_data` object throughout the codebase anyways it seemed beneficial to add the new type as I went along rather than having to do a completely separate pass on each object in another PR.
- Created the `NotificationDataAndCategory` type which maps a `NotificationCategory` to a specific notifications data type.
- Updated `emitNotifications` and all it's references to use `NotificationDataAndCategory` type (removes `object_id` references).
- Removed a raw query that subscribed users that were subscribed to `new-threads` to `new-comment-creation` on each new thread. If I hadn't deleted it, I would have had to refactor it to not use `object_id`. Zak cleared this deletion.
- Created the `mapNotificationsDataToSubscriptions` which as its name suggests, given some notification data will return the values from notification data that uniquely identify a subscription set.
- Added a validateSubscription validator on the Subscription Sequelize model. This validator ensures that only valid `NotificationCategories` are used. It also ensures that the proper data points are included for each category. For example, `new-thread-creation` subscriptions must always define a `chain_id`. This validation function mimics the database-level constraints but returns readable errors rather than unreadable PostgreSQL errors. The validation function is also important because it enforces Subscription validation for tests where `sequelize.sync` is used to create the database i.e. the database level constraints don't exist.
### Database
- Created a migration that does the following:
  - Some `new-thread` subscriptions were missing a `chain_id` so there is a query that updates these records.
  - Drops the `object_id` column from the `Subscriptions` table.
  - Created 5 new constraints on the `Subscriptions` table. These constraints are the same as those created in the `validateSubscription` function on the Subscription Sequelize model. The constraints ensure that dud subscriptions can't be created via raw SQL queries.

## Test Plan
[Notifications-QA][1]

## Deployment Plan

Normal deployment to CW.

## Other Considerations

I am aware that this is quite a large PR. That said, about 1.5k changes are purely test related and a large portion (500+ changes are purely structural -> structure of object/parameters changed). Furthermore, on top of the extensive test coverage added, I created the [Notifications-QA][1] document to ensure that all client-side functionality is maintained and regressions are minimized. All of the included changes are closely intertwined such that breaking the changes apart would be difficult and not worth the trouble. 

[1]: https://github.com/hicommonwealth/commonwealth/wiki/Notifications-QA
[2]: https://github.com/hicommonwealth/commonwealth/wiki/Subscription-Constraints

### CODE REVIEW

**Comment by timolegros** [comment]
PR is now ready for review but will remain in draft to avoid accidental merging until #4368 is merged.

**Comment by jnaviask** [comment]
Sadly need to bump this PR to release v0.6.4. Love the thorough QA plan, though. Should make it easy to release.

File: packages/commonwealth/client/scripts/controllers/server/notifications.ts [hunk 1]
```diff
@@ -82,29 +84,43 @@ class NotificationsController {
...
     );
+  }
+
+  public subscribe(data: SubUniqueData) {
+    const subscription = this.findNotificationSubscription(data);
     if (subscription) {
       return this.enableSubscriptions([subscription]);
     } else {
+      const requestData = Object.fromEntries(
+        Object.entries(data.options).map(([k, v]) => [
```
Comment (jnaviask) [inline packages/commonwealth/client/scripts/controllers/server/notifications.ts:106]: What does this regex do?

File: packages/commonwealth/server/migrations/20230807200113-rm-object-id.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+      // https://www.postgresql.org/docs/current/sql-altertable.html#SQL-ALTERTABLE-NOTES
+      // The NOT VALID option means that PostgreSQL will not scan the entire table to ensure the constraint holds for
+      // existing records. This is okay since we know that all existing records are valid. This hugely increases
+      // performance since the checks can be committed directly.
+      await queryInterface.sequelize.query(
+        `
+        ALTER TABLE "Subscriptions"
+            ADD CONSTRAINT "chk_chain_id_on_chain_event"
+                CHECK ((category_id <> 'chain-event') OR (category_id = 'chain-event' AND chain_id IS NOT NULL)) NOT VALID,
+            ADD CONSTRAINT "chk_snapshot_id_on_snapshot_proposal"
```
Comment (jnaviask) [inline packages/commonwealth/server/migrations/20230807200113-rm-object-id.js:30]: I worry about having undocumented constraints at the database level -- can these be reflected in the ORM? How will future developers know about these constraints?

**Review by jnaviask — COMMENTED** [review]
Quick, partial review here to surface some questions. Will need to take more time to thoroughly review later.

File: packages/commonwealth/server/migrations/20230807200113-rm-object-id.js (hunk 2 above)
Comment (timolegros) [reply packages/commonwealth/server/migrations/20230807200113-rm-object-id.js:30]: The exact same constraints are implemented as a validation function in the Sequelize model. Could add a wiki page describing how subscriptions work and in it include explanation of the constraints.

Comment (jnaviask) [reply packages/commonwealth/server/migrations/20230807200113-rm-object-id.js:30]: @timolegros could you add an inline comment to the sequelize model explaining how the validation function is reflected at the DB level? So we can keep the docs together.

File: packages/commonwealth/client/scripts/controllers/server/notifications.ts (hunk 1 above)
Comment (jnaviask) [reply packages/commonwealth/client/scripts/controllers/server/notifications.ts:106]: @timolegros please explicitly map the values if possible -- just for clarity's sake. I feel like this sort of automatic clever code can lead to mysterious bugs down the road. #4703 def worth pursuing in the future, or something similar.

Comment (timolegros) [reply packages/commonwealth/client/scripts/controllers/server/notifications.ts]: Done.

**Review by jnaviask — APPROVED** [review]

**Pull request merged**